src/function.js

```diff
@@ -93,12 +93,24 @@ Sk.builtin.checkIterable = function (arg) {
 };
 goog.exportSymbol("Sk.builtin.checkIterable", Sk.builtin.checkIterable);
 
-Sk.builtin.checkCallable = function (arg) {
-    if (typeof arg === "function") {
-        return (!(arg instanceof Sk.builtin.none) && (arg.ob$type !== undefined));
-    } else {
-        return ((arg.tp$call !== undefined) || (arg.__call__ !== undefined));
+Sk.builtin.checkCallable = function (obj) {
+    // takes care of builtin functions and methods, builtins
+    if (typeof obj === "function") {
+        return true;
     }
+    // takes care of python function, methods and lambdas
+    if (obj instanceof Sk.builtin.func) {
+        return true;
+    }
+    // takes care of instances of methods
+    if (obj instanceof Sk.builtin.method) {
+        return true;
+    }
+    // go up the prototype chain to see if the class has a __call__ method
+    if (Sk.abstr.lookupSpecial(obj, "__call__") !== undefined) {
```

Comment:
This doesn't work in the corner case where `__call__` is supplied by `__getattr__`. (That does work in real Python.)

It's a pretty corner-y case, and should probably be fixed all at once rather than in this PR, but I wanted to make a note.